src/middlewared/middlewared/plugins/reporting.py

```diff
@@ -0,0 +1,951 @@
+        return filter_list([i.__getstate__() for i in self.__rrds.values()], filters, options)
+
+    def __rquery_to_start_end(self, query):
+        unit = query.get('unit')
+        if unit:
+            verrors = ValidationErrors()
+            for i in ('start', 'end'):
+                if i in query:
+                    verrors.add(
+                        f'reporting_query.{i}',
+                        f'{i!r} should only be used if "unit" attribute is not provided.',
+                    )
+            verrors.check()
+        else:
+            if 'start' not in query:
+                unit = 'HOURLY'
+            else:
+                starttime = query['start']
+                endtime = query.get('end') or 'now'
+
+        if unit:
+            unit = unit[0].lower()
+            page = query['page']
+            starttime = f'end-{page + 1}{unit}'
+            if not page:
+                endtime = 'now'
+            else:
+                endtime = f'now-{page}{unit}'
+        return starttime, endtime
+
+    @accepts(
+        List('graphs', items=[
+            Dict(
+                'graph',
+                Str('name', required=True),
+                Str('identifier', default=None, null=True),
+            ),
+        ], empty=False),
+        Dict(
+            'reporting_query',
+            Str('unit', enum=[
+                'HOURLY', 'DAILY', 'WEEKLY', 'MONTHLY', 'YEARLY'
+            ]),
+            Int('page', default=0),
+            Str('start', empty=False),
+            Str('end', empty=False),
+            Bool('aggregate', default=True),
+            register=True,
+        )
+    )
+    def get_data(self, graphs, query):
+        """
+        Get reporting data for given graphs.
+
+        List of possible graphs can be retrieved using `reporting.graphs` call.
+
+        For the time period of the graph either `unit` and `page` OR `start` and `end` should be
+        used, not both.
+
+        `aggregate` will return aggregate available data for each graph (e.g. min, max, mean).
+
+        .. examples(websocket)::
+
+          Get graph data of "nfsstat" from the last hour.
+
+            :::javascript
+            {
+                "id": "6841f242-840a-11e6-a437-00e04d680384",
+                "msg": "method",
+                "method": "reporting.get_data",
+                "params": [
+                    [{"name": "nfsstat"}],
+                    {"unit": "HOURLY"},
+                ]
+            }
+
+        """
+        starttime, endtime = self.__rquery_to_start_end(query)
+        rv = []
+        for i in graphs:
+            try:
+                rrd = self.__rrds[i['name']]
+            except KeyError:
+                raise CallError(f'Graph {i["name"]!r} not found.', errno.ENOENT)
+            return rrd.export(i['identifier'], starttime, endtime, aggregate=query['aggregate'])
+        return rv
+
+    @private
+    @accepts(Ref('reporting_query'))
+    def get_all(self, query):
```

Comment:
I think maybe make passing empty list to `get_data` return data for all graphs? Or make `query` first argument and `graphs ` second and not specifying `graphs` do the same? I am bothered that two methods do essentially the same thing.

Reply:
get_all is a private method, I dont want people to be messing with it.

I also do not want it to be easy for people to get all graphs, thats a heavy call.

Reply:
Oh, I see, middleware does not call it, is it a debug method then?

Reply:
Yes, I can remove it if you think its not helpful

Reply:
No, let's keep it